django/thunderstore/social/permissions.py

```diff
@@ -0,0 +1,47 @@
+from abc import abstractmethod
+
+from django.conf import settings
+from rest_framework.exceptions import APIException, AuthenticationFailed
+from rest_framework.permissions import BasePermission
+
+
+class ImproperlyConfigured(APIException):
+    default_code = "improperly_configured"
+    default_detail = "Server is improperly configured"
+    status_code = 500
+
+
+class SharedSecretPermission(BasePermission):
+    """
+    Compares secret received in a header to one configured on the server.
+
+    Should be used for rudimentary access validation for endpoints that
+    are called by another one of our servers that has no access to user
+    credentials.
+    """
+
+    @property
+    @abstractmethod
+    def SHARED_SECRET(self) -> str:
+        """
+        Local value, against which the submitted value is checked.
+        """
+
+    def has_permission(self, request, view) -> bool:
+        if not self.SHARED_SECRET:
+            raise ImproperlyConfigured
+
+        header = request.META.get("HTTP_AUTHORIZATION")
```

Comment:
I wonder how does using the authorization header work in conjunction with the default DRF behavior. I believe authentication_classes rely on it and are checked before permission_classes, and I also believe some authentication classes will straight up error if an authorization header is present but not matching the format they expect (as would be the case with TS-Secret). Unsure if it would pose actual issues though.

One option to get around it for sure would be to use a custom header (e.g. `TS-Shared-Secret`), but honestly the authorization header seems the correct place to have this in to me.

I don't think this needs to be acted upon but wanted to note down my thoughts regardless, so feel free to mark resolved after reading unless you have other ideas.